009-strimzi-ui.md

```diff
@@ -0,0 +1,262 @@
+
+## Session management and Authentication
+
+This section describes how a UI deployed as a part of Strimzi could be secured. [The following is based on the work done as a part of the security proposal, and how the UI would/could utilise the available security mechanisms.](https://github.com/strimzi/proposals/pull/13). The exact implementation, as well as any other required discussion around security I would suggest is had as a part of the UI's development/delivery to Strimzi.
+
+### Aim:
+
+Provide a mechanism for UI users to log into Strimzi and make authorized requests to Kafka.
+
+### Overview:
+
+UX would be impacted if a user had to provide credentials for every request made by the UI, so instead a user should log in once, and credentials are then automatically appended to requests. This session should have a maximum lifetime, at which point a user will need to re-authenticate. In addition, sessions should expire due to inactivity. A user should also be able to log out of a UI to allow for user switching.
+
+This session must be shared between HTTP and WebSocket traffic – as the UI will be executing graphql queries, mutations and subscriptions using the Apollo graphql stack.
+
+### Proposed architecture:
+
+![UI session component architecture](./images/009-session-architecture.png)
+
+- Admin – Graphql server, supporting HTTP and websocket connections [admin server proposal](https://github.com/strimzi/proposals/pull/9)
+- Express – UI server, handling sessions for the client
+- Client – Browser running the strimzi UI, executing graphql queries/subscriptions
+- Kafka – brokers (connected to via admin clients)
+- Auth – external auth provider, oauth for Kafka (and possible oauth dance for UI)
+
+### Proposed technology:
+
+The UI is being hosted by an Express server. Express has session middleware - https://www.npmjs.com/package/express-session - that can create and persist server-side sessions, using a cookie as a key to hydrate a session into the incoming express request. Proposed session store is a `redis` container as it's a lightweight key/value (and the session will just be storing a token value).
+
+Authentication in node/Express can be achieved through by http://www.passportjs.org/ which provides a large collection of “strategies” for authenticating a user. For purposes of the proposal, the below diagrams show an oidc flows for a UI oauth dance, although other authentication mechanisms could be used.
+
+### Sequence diagrams:
+
+Note – the log in flow is assuming an oauth dance to retrieve a user token. Depending on the passport strategy and kafka auth mechanism, the UI may need to render its own basic login page.
+
+#### http flow with valid session
+
+![UI http flow with valid session](./images/009-http-valid-session.svg)
```

Comment:
as reported in #13 an alternative is to check the JWT token locally without introspection to the IDP.

Reply:
I will add this for completness also